crates/aiken-project/src/blueprint/validator.rs

```diff
@@ -0,0 +1,631 @@
+    }
+}
+
+impl Display for Validator {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        let s = serde_json::to_string_pretty(self).map_err(|_| fmt::Error)?;
+        f.write_str(&s)
+    }
+}
+
+impl Validator {
+    pub fn from_checked_module(
+        modules: &CheckedModules,
+        generator: &mut CodeGenerator,
+        validator: &CheckedModule,
+        def: &TypedFunction,
+    ) -> Result<Validator, Error> {
+        let purpose: Purpose = def.name.clone().into();
+
+        assert_return_bool(validator, def)?;
+        assert_min_arity(validator, def, purpose.min_arity())?;
+
+        let mut args = def.arguments.iter().rev();
+        let (_, redeemer, datum) = (args.next(), args.next().unwrap(), args.next());
+
+        Ok(Validator {
+            title: validator.name.clone(),
+            description: None,
+            purpose,
+            datum: datum
+                .map(|datum| {
+                    Annotated::from_type(modules.into(), &datum.tipo, &HashMap::new()).map_err(
+                        |error| Error::Schema {
+                            error,
+                            location: datum.location,
+                            source_code: NamedSource::new(
+                                validator.input_path.display().to_string(),
+                                validator.code.clone(),
+                            ),
+                        },
+                    )
+                })
+                .transpose()?,
+            redeemer: Annotated::from_type(modules.into(), &redeemer.tipo, &HashMap::new())
+                .map_err(|error| Error::Schema {
+                    error,
+                    location: redeemer.location,
+                    source_code: NamedSource::new(
+                        validator.input_path.display().to_string(),
+                        validator.code.clone(),
+                    ),
+                })?,
+            program: generator
+                .generate(&def.body, &def.arguments, true)
+                .try_into()
+                .unwrap(),
+        })
+    }
+}
+
+impl Purpose {
+    pub fn min_arity(&self) -> u8 {
+        match self {
+            Purpose::Spend => 3,
+            Purpose::Mint | Purpose::Withdraw | Purpose::Publish => 2,
+        }
+    }
+}
+
+impl Display for Purpose {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.write_str(match self {
+            Purpose::Spend => "spend",
+            Purpose::Mint => "mint",
+            Purpose::Withdraw => "withdraw",
+            Purpose::Publish => "publish",
+        })
+    }
+}
+
+impl From<String> for Purpose {
+    fn from(purpose: String) -> Purpose {
+        match &purpose[..] {
+            "spend" => Purpose::Spend,
+            "mint" => Purpose::Mint,
+            "withdraw" => Purpose::Withdraw,
+            "publish" => Purpose::Publish,
+            unexpected => panic!("Can't turn '{}' into any Purpose", unexpected),
```

Comment:
Depending on whether on not we actually expect this to happen we could use TryFrom